### ISSUE TITLE
Blender: export multiple animations

### ISSUE THREAD

**Issue by arturitu** [mrdoob/three.js#6326]
I´m trying to export multiple animations with the new Blender exporter but it seems not work properly (I'm testing with a project similar to the http://threejs.org/examples/webgl_animation_skinning_blending.html example ) but with the old exporter works well (in this case removing the armature previously)

**Comment by titansoftime** [comment]
You must set both the rig and the skeletal export properties to "rest". See #5887

**Comment by repsac** [comment]
Correct. The export properties must match whether the rig is set to Pose Position or Rest Position. Depending on how your asset is setup (some people need to do Rest and some need to do Pose) will determine which export results in the correct data so there is a bit of trial and error. I really wish we didn't need to these 2 separate animation export options but it seems to be a necessity (at this time).

I experimented with setting the pose position interactively based on the exporter settings, but this didn't seem to work correctly. The code is still in place but commented out until I can find the patience to look at it again (kind of taking a break during r71). 

Additionally; uncheck the `Indent` export option to reduce the file size.

**Comment by tmunz** [comment]
Thanks, I will have a look on that the next days

**Comment by satori99** [comment]
> I experimented with setting the pose position interactively based on the exporter settings, but this didn't seem to work correctly.

Were you trying to set `armature.pose_position = 'REST'` in the addon code? I came across a problem doing this in a blender script recently. The armature is not actually put in a rest pose until you also call `context.scene.update()`

I was confused because setting the `pose_position` in the python console works without an explicit scene update.

**Comment by repsac** [comment]
@satori99 that was exactly the problem I ran into.

**Comment by satori99** [comment]
I wasted half a day working that out.

I'd do a PR if I was more familiar with the addon code. But it sounds like it's still experimental code anyway.

**Comment by arturitu** [comment]
I tried all combinations but I have several problems exporting my character, I'll try to explain:

* Here is the project with the OLD classes working ( r68 ) > http://unboring.net/lab/6326/hero_68/
The source ( includes .blend file ) > http://unboring.net/lab/6326/hero_68.zip

![screenshot 2015-04-04 09 03 59](https://cloud.githubusercontent.com/assets/5744112/6991798/94a3bb1a-daa9-11e4-87ec-1334cc670232.png)

* Here is the project with the new classes NOT working ( r71 ) > http://unboring.net/lab/6326/hero_71/
The source ( includes .blend file ) > http://unboring.net/lab/6326/hero_71.zip

 * When I export ["Pose / Pose" JSON file] (http://unboring.net/lab/6326/hero_71/obj/hero_pose_pose.json) ( Armature in Pose mode / Export animation ( skeletal ): Pose ) exported shows 38 bones ( **some of the are IK controllers and maybe is the problem** )
and the parent value is correct but the result is this:

![screenshot 2015-04-04 09 02 01](https://cloud.githubusercontent.com/assets/5744112/6991808/751959b6-daaa-11e4-9a7e-00b7154916e0.png)

 * When I export ["Pose / Rest"  JSON file] (http://unboring.net/lab/6326/hero_71/obj/hero_pose_rest.json) only includes 28 bones and the parent value of the bones is null in most and the result is an error.

 * When I export ["Rest / Pose" JSON file] (http://unboring.net/lab/6326/hero_71/obj/hero_rest_pose.json) exported shows 38 bones and the parent value is correct but the result is this ( is not animated ):

![screenshot 2015-04-04 09 15 11](https://cloud.githubusercontent.com/assets/5744112/6991817/660035c0-daab-11e4-8e13-034edc6a6052.png)

 * When I export ["Rest / Rest"  JSON file] (http://unboring.net/lab/6326/hero_71/obj/hero_rest_rest.json) only includes 28 bones and the parent value of the bones is null in most and the result is an error but the JSON looks more efficient because the animations keys only includes keyframes.

**I don't know if is problem of my blend file, of the BlendCharacter that is not ready for the new exporter class or the exporter is not ready to work with models that has bones that are used as controllers.**

**Comment by repsac** [comment]
Downloaded the file. Will look at it this week (been avoiding the computer on the weekends).

**Comment by repsac** [comment]
Mixing pose/rest isn't recommended. Looks like Pose/Pose is what you need (r67 to r69 used pose armatures) so there may be something wrong when the pose logic was migrated to r70.

**Comment by repsac** [comment]
Running into lots of issues trying to make sense here. I keep legacy versions of `three` installed so I can quickly swing symlinks to other versions of the exporter. My addons setup is linked to the `r68` exporter:
```
10:44 ~/scripts/BLENDER_USER_SCRIPTS/addons
$ pwd
/Users/ed/scripts/BLENDER_USER_SCRIPTS/addons
10:44 ~/scripts/BLENDER_USER_SCRIPTS/addons
$ ll
total 24
lrwxr-xr-x  1 ed  staff    98 Apr 13 10:27 io_mesh_threejs -> ../../../work/www/threejs/three.js-r68/utils/exporters/blender/2.65/scripts/addons/io_mesh_threejs
lrwxr-xr-x  1 ed  staff    68 Mar  9 20:04 io_three -> ../../../work/repos/three.js/utils/exporters/blender/addons/io_three
-rw-r--r--  1 ed  staff  2934 Nov 29  2013 theme_export.py
```
![screen shot 2015-04-13 at 10 45 59 am](https://cloud.githubusercontent.com/assets/1109289/7121590/683281c4-e1ca-11e4-982d-374cc0f7d2df.png)

But when I exported I got a `hero.js` file that did not match the working example I downloaded.
http://pastebin.com/PT00GfqK

Not being able to recreate a correct export makes it harder to figure out why the other exports aren't correct.

**Comment by arturitu** [comment]
@repsac this is because with the old exporter, you need to remove the armature modifier before to export. If I export with the armature I obtain 1184 normals too, but the correct .json has 950 (without armature modifier)

Not ready to export
![image](https://cloud.githubusercontent.com/assets/5744112/7122052/f31c2be8-e218-11e4-8215-7ff02cbbe16e.png)

Ready to export
![image](https://cloud.githubusercontent.com/assets/5744112/7122065/0390b688-e219-11e4-88f2-1b91bd7fe262.png)

Could that be the problem?

**Comment by repsac** [comment]
OK, that clears up what you meant by:

> in this case removing the armature previously

So now the normals match but the bones and animation arrays are still different.

**Comment by arturitu** [comment]
It's very strange but only works well when you export (without armature modifier and) with the armature selected like this >

![image](https://cloud.githubusercontent.com/assets/5744112/7124805/425fca30-e22c-11e4-8506-1ff89733c709.png)

If you export without nothing or the mesh selected don't export correctly.

These are the options of the exporter:

![image](https://cloud.githubusercontent.com/assets/5744112/7124788/1d234576-e22c-11e4-8d0d-c24a9cea63e3.png)

**Comment by laetitiakami** [comment]
About the bones having "null" parent, check if "use_deform" property is checked for the parent bones. Uncheck it and it should work. After reading the code I believe such bones are ignored. I am also trying to make animations work with the last exporter.

**Comment by fredleefarr** [comment]
I was having similar issues with multiple animations exported from Blender. What would be the recommended approach - wait for further r71 development, or use the suggestions in this thread?

**Comment by repsac** [comment]
I don't have any suggestions. Trying to make sense on why someone would need to remove the armature modifier. Not saying that @arturitu is wrong in his work flow (doing what he has to do to get his animations) but it is not easy to understand why it is necessary. If people were consistently having issues across the board it would be one thing but that doesn't seem to be the case. Some things may be directly related to the exporter, but other things may be on how rigs are setup in Blender. We never established an specification for how to rig for `three` animations but also because I don't think anyone really knows what those should be.

This has really been the only real issue to come up for r71 but it has been one of the most difficult I have addressed.

**Comment by arturitu** [comment]
I'm a noob rigging, but when I did this character I tried to follow tutorials like this ( http://cgi.tutsplus.com/tutorials/building-a-basic-low-poly-character-rig-in-blender--cg-16955 ) about how to rig with Blender. And I removed the armature because was the only way to worked for me.

If we could establish an specification for how to rig will be great (if that way we can use the new exporter with multiple animations). It looks more efficient because when use Rest/Rest seems that only exports the keyframes for each bone, not every frame of every animation.

**Comment by repsac** [comment]
> And I removed the armature because was the only way to worked for me.

Sometimes you do what you got to do. 

> establish an specification for how to rig will be great

oh believe me I would love to get this figured out and documented, would save on so much time by addressing less issues

> only exports the keyframes for each bone, not every frame of every animation.

`Rest` and `Pose` exports were written by different people (if I remember correctly). `Rest` exports had the added optimization that would avoid creating an entry on a keyframe if values hadn't change. I am aware that this optimization wasn't done for `Pose`. It's on the "todo" list.

**Comment by fredleefarr** [comment]
> If we could establish an specification for how to rig will be great

+1 from me

**Comment by GoroMatsumoto** [comment]
Hi. This is my workaround with r70.
I export json for mesh, and animations for each.
Exporting an animation is enable for "active" action only.
So you have to export all animations for each.
This is my setting:
![blenderexportsetting](https://cloud.githubusercontent.com/assets/1933467/7339552/87ee2c8c-ecad-11e4-818a-ea01cbf4cf30.jpg)

need sleep...
I will continue...

Cheers.

**Comment by manthrax** [comment]
There is a possible workaround/fix for some of the multi-animation export problems here... #6050
